opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -350,6 +350,80 @@ std::vector<std::vector<int>> communicateExportedWells(
     return result;
 }
 
+std::vector<std::vector<int>> nonblockingCommunicateExportedWells(
+    const std::vector<std::vector<std::vector<int>>>& exportedWells,
+    const Dune::cpgrid::CpGridDataTraits::Communication& cc,
+    int root)
+{
+    // send data from root
+    std::vector<std::vector<int>> result;
+    if (cc.rank()==root)
+    {
+        assert((int)exportedWells.size()==cc.size());
+        assert((int)exportedWells[root].size()==0);
+        // check send success for each well
+        std::vector<MPI_Request> requestSize(cc.size());
+        std::vector<std::vector<MPI_Request>> requestData(cc.size());
+
+        // number of wells and their sizes on each rank
+        // equal to exportedWells***.size() but we can not send temporary
+        std::vector<int> nrWells(cc.size());
+        std::vector<std::vector<int>> wellSizes(cc.size());
+        for (int i=0; i<cc.size(); ++i)
+        {
+            nrWells[i]=exportedWells[i].size();
+            requestData[i].resize(2*nrWells[i]);
+            wellSizes[i].reserve(nrWells[i]);
+            for (const auto& ewi : exportedWells[i])
+            {
+                wellSizes[i].push_back(ewi.size());
+            }
+        }
+
+        for (int i=0; i<cc.size(); ++i)
+        {
+            if (i!=root)
+            {
+                int tag = 37; // a random number
+                MPI_Isend(&nrWells[i], 1, MPI_INT, i, tag++, cc, &requestSize[i]);
+                for (int j=0; j<(int)exportedWells[i].size(); ++j)
+                {
+                    assert(j+1<(int)requestData[i].size());
+                    MPI_Isend(&wellSizes[i][j], 1, MPI_INT, i, tag++, cc, &requestData[i][2*j]);
+                    MPI_Isend(exportedWells[i][j].data(), wellSizes[i][j], MPI_INT, i, tag++, cc, &requestData[i][2*j+1]);
```

Comment:
Just a note/remark:

This is probably not performance critical and would be more work, but generally one could merge these two messages into one using an extra buffer.

For that one would need to query the message size beforre receiving using MPI_Test and MPI_Status::count. Then allocate buffers and receive.

Reply:
I do not see how they could be merged, since wellSizes are received in the first message and used for the size in the second message. But I found different simplification: wellSizes can be communicated as a vector.